lib/handler/cache-handler.js

```diff
@@ -0,0 +1,364 @@
+    }
+
+    this.#opts = opts
+    this.#req = req
+    this.#handler = handler
+  }
+
+  /**
+   * @see {DispatchHandlers.onHeaders}
+   *
+   * @param {number} statusCode
+   * @param {Buffer[]} rawHeaders
+   * @param {() => void} resume
+   * @param {string} statusMessage
+   * @returns {boolean}
+   */
+  onHeaders (
+    statusCode,
+    rawHeaders,
+    resume,
+    statusMessage
+  ) {
+    const downstreamOnHeaders = () => this.#handler.onHeaders(
+      statusCode,
+      rawHeaders,
+      resume,
+      statusMessage
+    )
+
+    if (
+      UNSAFE_METHODS.includes(this.#req.method) &&
+      statusCode >= 200 &&
+      statusCode <= 399
+    ) {
+      // https://www.rfc-editor.org/rfc/rfc9111.html#name-invalidating-stored-respons
+      // Try/catch for if it's synchronous
+      try {
+        const result = this.#opts.store.deleteByOrigin(this.#req.origin)
+        if (
+          result &&
+          typeof result.catch === 'function' &&
+          typeof this.#handler.onError === 'function'
+        ) {
+          // Fail silently
+          result.catch(_ => {})
+        }
+      } catch (err) {
+        // Fail silently
+      }
+
+      return downstreamOnHeaders()
+    }
+
+    const headers = util.parseHeaders(rawHeaders)
+
+    const cacheControlHeader = headers['cache-control']
+    const contentLengthHeader = headers['content-length']
+
+    if (!cacheControlHeader || !contentLengthHeader || this.#opts.store.isFull) {
+      // Don't have the headers we need, can't cache
+      return downstreamOnHeaders()
+    }
+
+    const contentLength = Number(contentLengthHeader)
+    if (!Number.isInteger(contentLength)) {
+      return downstreamOnHeaders()
+    }
+
+    const cacheControlDirectives = parseCacheControlHeader(cacheControlHeader)
+    if (!canCacheResponse(statusCode, headers, cacheControlDirectives)) {
+      return downstreamOnHeaders()
+    }
+
+    const now = Date.now()
+    const staleAt = determineStaleAt(now, headers, cacheControlDirectives)
+    if (staleAt) {
+      const varyDirectives = headers.vary
+        ? parseVaryHeader(headers.vary, this.#req.headers)
+        : undefined
+      const deleteAt = determineDeleteAt(now, cacheControlDirectives, staleAt)
+
+      const strippedHeaders = stripNecessaryHeaders(
+        rawHeaders,
+        headers,
+        cacheControlDirectives
+      )
+
+      this.#writeStream = this.#opts.store.createWriteStream(this.#req, {
+        statusCode,
+        statusMessage,
+        rawHeaders: strippedHeaders,
+        vary: varyDirectives,
+        cachedAt: now,
+        staleAt,
+        deleteAt
+      })
+
+      if (this.#writeStream) {
+        this.#writeStream.on('drain', resume)
+        this.#writeStream.on('error', () => {
+          this.#writeStream = undefined
+          resume()
+        })
+      }
+    }
+
+    if (typeof this.#handler.onHeaders === 'function') {
+      return downstreamOnHeaders()
+    }
+  }
```

Comment:
I guess also here return false and not undefined?!
```suggestion
    }
    return false
  }
```